src/core/expression/qgsexpressionfunction.cpp

```diff
@@ -4232,3 +4286,96 @@ const QList<QgsExpressionFunction *> &QgsExpression::Functions()
   }
   return sFunctions;
 }
+
+QgsWithVariableExpressionFunction::QgsWithVariableExpressionFunction()
+  : QgsExpressionFunction( QStringLiteral( "with_variable" ), 3, QCoreApplication::tr( "General" ) )
+{
+
+}
+
+bool QgsWithVariableExpressionFunction::isStatic( const QgsExpressionNodeFunction *node, QgsExpression *parent, const QgsExpressionContext *context ) const
+{
+  bool isStatic = false;
+
+  QgsExpressionNode::NodeList *args = node->args();
+
+  if ( args->count() < 3 )
+    return false;
+
+  if ( args->at( 0 )->isStatic( parent, context ) && args->at( 1 )->isStatic( parent, context ) )
+  {
+    QVariant name = args->at( 0 )->eval( parent, context );
+    QVariant value = args->at( 1 )->eval( parent, context );
+
+    QgsExpressionContextScope *scope = new QgsExpressionContextScope();
+    scope->setVariable( name.toString(), value );
+
+    QgsExpressionContext *updatedContext = const_cast<QgsExpressionContext *>( context );
```

Comment:
Not sure what the logic is here -can you add some comments to explain why this temporary scope is required?